src/node.rs

```diff
@@ -114,16 +126,263 @@ impl InMemoryNodeInner {
+        let estimated_fee_scale_factor = 1.3;
+        let tx_body_gas_limit = cmp::min(
+            MAX_L2_TX_GAS_LIMIT as u32,
+            (upper_bound as f32 * estimated_fee_scale_factor) as u32,
+        );
+        let suggested_gas_limit = tx_body_gas_limit + gas_for_bytecodes_pubdata;
+
+        let estimate_gas_result = self.estimate_gas_step(
+            l2_tx.clone(),
+            gas_per_pubdata_byte,
+            suggested_gas_limit,
+            l1_gas_price,
+            base_fee,
+        );
+
+        match estimate_gas_result {
+            Err(_) => Err(into_jsrpc_error(Web3Error::SubmitTransactionError(
+                "Transaction is unexecutable".into(),
+                Default::default(),
+            ))),
+            Ok(_) => {
+                let overhead: u32 = derive_gas_estimation_overhead(
+                    suggested_gas_limit,
+                    gas_per_pubdata_byte as u32,
+                    tx.encoding_len(),
+                );
+
+                let full_gas_limit =
+                    match tx_body_gas_limit.overflowing_add(gas_for_bytecodes_pubdata + overhead) {
+                        (value, false) => value,
+                        (_, true) => {
+                            return Err(into_jsrpc_error(Web3Error::SubmitTransactionError(
+                                "exceeds block gas limit".into(),
+                                Default::default(),
+                            )))
+                        }
+                    };
+
+                let fee = Fee {
+                    max_fee_per_gas: base_fee.into(),
+                    max_priority_fee_per_gas: 0u32.into(),
+                    gas_limit: full_gas_limit.into(),
+                    gas_per_pubdata_limit: gas_per_pubdata_byte.into(),
+                };
+                Ok(fee)
+            }
+        }
+    }
+
+    /// Runs fee estimation against a sandbox vm with the given gas_limit.
+    fn estimate_gas_step(
+        &self,
+        mut l2_tx: L2Tx,
+        gas_per_pubdata_byte: u64,
+        tx_gas_limit: u32,
+        l1_gas_price: u64,
+        base_fee: u64,
+    ) -> Result<VmBlockResult, TxRevertReason> {
+        let execution_mode = TxExecutionMode::EstimateFee {
+            missed_storage_invocation_limit: 1000000,
+        };
+
+        let tx: Transaction = l2_tx.clone().into();
+        let l1_gas_price =
+            adjust_l1_gas_price_for_tx(l1_gas_price, L2_GAS_PRICE, tx.gas_per_pubdata_byte_limit());
+
+        // Set gas_limit for transaction
+        let gas_limit_with_overhead = tx_gas_limit
+            + derive_gas_estimation_overhead(
+                tx_gas_limit,
+                gas_per_pubdata_byte as u32,
+                tx.encoding_len(),
+            );
+        l2_tx.common_data.fee.gas_limit = gas_limit_with_overhead.into();
+
+        let fork_storage_copy = self.fork_storage.clone();
+        let mut storage_view = StorageView::new(&fork_storage_copy);
```

Comment:
No need to `clone` here. We want to be cautious in the usage of `clone` as it can be a costly operation. Since `StorageView` requires a reference we can adjust to:
```
let fork_storage_copy = &self.fork_storage;
let mut storage_view = StorageView::new(fork_storage_copy);
```

Reply:
Agreed, good catch. I need to double check that a full clone isn't necessary in the sandbox vm. Will update here.

Reply:
Confirmed the Greeter contract test still works without a full clone. Just still worried since we're updating the underlying values, but something to watch out for.